src/staking/RewardTracker.sol

```diff
@@ -0,0 +1,257 @@
+        require(isVault[address(vault)], "vault not whitelisted");
+        uint256 beforeBalance = rewardToken.balanceOf(address(this));
+        vault.redeem(vault.balanceOf(address(this)), address(this), address(this));
+        uint256 afterBalance = rewardToken.balanceOf(address(this));
+        _notifyRewardAmount(afterBalance - beforeBalance);
+    }
+
+    /// @notice Lets an admin add a vault for collecting fees from.
+    function addVault(address vault) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(ERC4626(vault).asset() == rewardToken, "only WETH assets");
+        isVault[vault] = true;
+        emit VaultAdded(vault);
+    }
+
+    function _earned(address account, uint256 accountBalance, uint256 rewardPerToken_, uint256 accountRewards)
+        internal
+        view
+        returns (uint256)
+    {
+        return accountBalance.mulDivDown(rewardPerToken_ - userRewardPerTokenPaid[account], PRECISION) + accountRewards;
+    }
+
+    function _rewardPerToken(uint256 totalSupply_, uint256 lastTimeRewardApplicable_, uint256 rewardRate_)
+        internal
+        view
+        returns (uint256)
+    {
+        if (totalSupply_ == 0) {
+            return rewardPerTokenStored;
+        }
+        return rewardPerTokenStored
+            + rewardRate_.mulDivDown((lastTimeRewardApplicable_ - lastUpdateTime) * PRECISION, totalSupply_);
+    }
+
+    function _updateReward(address account) internal override {
+        // storage loads
+        uint256 accountBalance = balanceOf[account] + multiplierPointsOf[account];
+        uint64 lastTimeRewardApplicable_ = lastTimeRewardApplicable();
+        uint256 totalSupply_ = totalSupply + totalBonus;
+        uint256 rewardPerToken_ = _rewardPerToken(totalSupply_, lastTimeRewardApplicable_, rewardRate);
+
+        // accrue rewards
+        rewardPerTokenStored = rewardPerToken_;
+        lastUpdateTime = lastTimeRewardApplicable_;
+        rewards[account] = _earned(account, accountBalance, rewardPerToken_, rewards[account]);
+        userRewardPerTokenPaid[account] = rewardPerToken_;
+    }
+
+    function transfer(address to, uint256 amount) public override returns (bool) {
+        _updateReward(msg.sender);
+        _updateReward(to);
+        _updateBonus(msg.sender);
+        _updateBonus(to);
+        _burnMultiplierPoints(amount, msg.sender);
+        return super.transfer(to, amount);
+    }
+
+    function transferFrom(address from, address to, uint256 amount) public override returns (bool) {
+        _updateReward(from);
+        _updateReward(to);
+        _updateBonus(from);
+        _updateBonus(to);
+        _burnMultiplierPoints(amount, from);
+        return super.transferFrom(from, to, amount);
+    }
+
+    // burn multiplier points
+    function _burnMultiplierPoints(uint256 amount, address sender) internal {
+        uint256 bonus_ = multiplierPointsOf[sender];
+
+        // if the sender has bonus points
+        if (bonus_ > 0) {
```

Comment:
just a suggestion to change to `if(bonus == 0) return;` but that's nitpicking

Reply:
fixed